system/jlib/jtrace.hpp

```diff
@@ -55,6 +231,145 @@ class jlib_decl LogTrace
+ */
+namespace HPCCSemanticConventions
+{
+static constexpr const char *kGLOBALIDHTTPHeader = "HPCC-Global-Id";
+static constexpr const char *kCallerIdHTTPHeader = "HPCC-Caller-Id";
+}
+
+// github copilot generated comment:
+// This class provides a high-level interface for managing tracing and profiling information in the HPCC Systems platform.
+// The TraceManager class is responsible for creating and managing instances of the Tracer class, which is used to instrument code for tracing and profiling. The Tracer class provides methods for starting and ending spans, adding attributes to spans, and propagating trace context across different services and systems.
+// Overall, the TraceManager class provides a convenient and flexible way to instrument code for tracing and profiling in the HPCC Systems platform. By using the Tracer class and the TraceManager class, developers can gain insight into the performance and behavior of their applications and diagnose issues in distributed systems.
+class TraceManager
+{
+private:
+    //Used as the opentel trace name, refered to as name of library being instrumented
+    const char * tracerName = nullptr;
+
+public:
+    TraceManager(const char * moduleName)
+    {
+        //InitModuleObjects();
+        initTracer(); //Still not sure where this should be done,
+                      //but it needs to be done once per process before any tracers/spans are created
+        tracerName = moduleName; //1 tracer for each module/library being instrumented
+    };
+
+    TraceManager(const std::string & moduleName)
+    {
+        TraceManager(moduleName.c_str());
+    };
+
+    ~TraceManager() {};
+
+    static void initTracer();
+    static void cleanupTracer();
+
+    //convenience non-static method to get the default tracer, uses stored tracer/module name
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> getTracer();
+
+    //convenience Static method to get the tracer for the provided module name
+    static opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> getTracer(std::string moduleName);
+
+    //Extracts parent contex from the carrier's headers and returns it as callerSpanId
+    static void getCallerSpanId(context::propagation::TextMapCarrier &carrier, std::string & callerSpanId);
+    static void getParentSpanId(std::map<std::string, std::string> requestHeaders, std::string & callerSpanId);
+    static void getParentSpanId(const HttpTextMapCarrier<std::map<std::string, std::string>> carrier, std::string & callerSpanId);
+    static void getParentContext(std::map<std::string, std::string>& request_headers, opentel_trace::StartSpanOptions & options);
+    
+    //Get the parentSpan from ANY Carrier implementation
+    // HPCCHttpTextMapCarrier<IProperties> , HttpTextMapCarrier<http_client::Headers>, etc
+    // The carrier must implement the TextMapCarrier interface
+    // and it will determine the header which contains the parent span
+    template<class CARRIER>
```

Comment:
Again the use of templates suggests a lack of design.   The HPCC codebase tends to use interfaces rather than templates classes. 
templates are generally reserved for 
- simple integer types
- generic general purpose container classes (e.g. hash table)
- rarely implementing classes on top of other base classes.